File: cpp_src/BoundaryGenerationService.cpp

#include <iostream>
#include <string>
#include <exception>
#include <map>
#include "BoundaryGenerationService.h"

using namespace std;



BoundaryGenerationService::BoundaryGenerationService(int rows, int cols, vector<vector<double>> &mesh, vector<vector<bool>> &fixed_indices): rows(rows), cols(cols), mesh(mesh), fixed_indices(fixed_indices){}




bool BoundaryGenerationService::place_rectangle_potential_boundary(int x, int y, int width, int height, double potential){
    /*
      Draw a boundary rectangle with fixed values inside and on the edges,
      its edges parallel to the grid-lines

    */
   

    if( x + width > cols || y + height > rows ){
        return false;
    }

   
   for(int i = 0; i < width; i++){
        for(int j = 0; j < height ; j++){

    	    mesh[x + i][y + j] = potential;
	        fixed_indices[x + i][y + j] = true;

        }
    }

    return true;

}


bool BoundaryGenerationService::place_circle_potential_boundary(int centre_x, int centre_y, int radius, circle_boundary_params params){


    vector<vector<bool>> has_not_been_filled(rows, vector<bool>(cols, true));

    //Implementation of Bresenham's Algorithm for boundary
    try 
    {

        int x_pos = 0, y_pos = radius;
        int decision_param = 3 - (2 * radius);

        complete_octants(centre_x, centre_y, x_pos, y_pos, params.boundary_potential, params.boundary_fixed, has_not_been_filled);

        while( y_pos >= x_pos ) {

            ++x_pos;

            if (decision_param > 0){

                --y_pos;
                decision_param = decision_param + 4 * (x_pos - y_pos) + 10;

            } else {

                decision_param = decision_param + 4 * x_pos + 6;

            }
        
            complete_octants(centre_x, centre_y, x_pos, y_pos, params.boundary_potential, params.boundary_fixed, has_not_been_filled);

        }
    }
    catch(exception& e) 
    {
        return false;
    }

    //Fill the interior
    if(params.internal_potential != 0 && params.internal_fixed != false)
        fill_tool(centre_x, centre_y,params.internal_potential, params.internal_fixed, has_not_been_filled);

    //Fill the exterior
    fill_exterior(params.external_potential, params.external_potential, has_not_been_filled);
    
    return true;
}

void BoundaryGenerationService::complete_octants(int centre_x, int centre_y, int x, int y, double potential, bool boundary_fixed, vector<vector<bool>> &has_not_been_filled){

    try
    {
        mesh[centre_x + x][ centre_y + y] = potential;
        mesh[centre_x - x][ centre_y + y] = potential;
        mesh[centre_x + x][ centre_y - y] = potential;
        mesh[centre_x - x][ centre_y - y] = potential;
        mesh[centre_x + y][ centre_y + x] = potential;
        mesh[centre_x - y][ centre_y + x] = potential;
        mesh[centre_x + y][ centre_y - x] = potential;
        mesh[centre_x - y][ centre_y - x] = potential;

        fixed_indices[centre_x + x][ centre_y + y] = boundary_fixed;
        fixed_indices[centre_x - x][ centre_y + y] = boundary_fixed;
        fixed_indices[centre_x + x][ centre_y - y] = boundary_fixed;
        fixed_indices[centre_x - x][ centre_y - y] = boundary_fixed;
        fixed_indices[centre_x + y][ centre_y + x] = boundary_fixed;
        fixed_indices[centre_x - y][ centre_y + x] = boundary_fixed;
        fixed_indices[centre_x + y][ centre_y - x] = boundary_fixed;
        fixed_indices[centre_x - y][ centre_y - x] = boundary_fixed;

        has_not_been_filled[centre_x + x][ centre_y + y] = false;
        has_not_been_filled[centre_x - x][ centre_y + y] = false;
        has_not_been_filled[centre_x + x][ centre_y - y] = false;
        has_not_been_filled[centre_x - x][ centre_y - y] = false;
        has_not_been_filled[centre_x + y][ centre_y + x] = false;
        has_not_been_filled[centre_x - y][ centre_y + x] = false;
        has_not_been_filled[centre_x + y][ centre_y - x] = false;
        has_not_been_filled[centre_x - y][ centre_y - x] = false;

    }
    catch(exception& e)
    {
        throw e;

    }
}



//Recursively fill the whatever part of the mesh that hasn't been set yet
//For circle, pass a point inside to fill its interior 
void BoundaryGenerationService::fill_tool(int x, int y, double potential, bool fixed, vector<vector<bool>> &has_not_been_filled){

    if(has_not_been_filled[x][y]){

        mesh[x][y] = potential;
        fixed_indices[x][y] = fixed;
        has_not_been_filled[x][y] = false;

        fill_tool(x+1, y, potential, fixed, has_not_been_filled);
        fill_tool(x-1, y, potential, fixed, has_not_been_filled);
        fill_tool(x, y+1, potential, fixed, has_not_been_filled);
        fill_tool(x, y-1, potential, fixed, has_not_been_filled);

    }
    else {
        return;
    }


}



void BoundaryGenerationService::fill_exterior(double exterior_potential, bool exterior_fixed,  vector<vector<bool>> &has_not_been_filled){

    //Simple way is just to check all the boundary points

    //Check top edge
    for(int i = 0; i != 50; ++i){
        if (has_not_been_filled[i][0])
        {
            fill_tool(i, 0, exterior_potential, exterior_fixed, has_not_been_filled);
        }   
    }

    //check right sie
    for(int j = 1; j != 50; ++j){
                if (has_not_been_filled[rows-1][j])
        {
            fill_tool(rows-1, j, exterior_potential, exterior_fixed, has_not_been_filled);
        }   
    }

    //check bottom side
    for(int i = 0; i != 49; ++i){
        if (has_not_been_filled[i][cols-1])
        {
            fill_tool(i, cols-1, exterior_potential, exterior_fixed, has_not_been_filled);
        }   
    }


    //Check the left side
    for(int j = 1; j != 49; ++j){
                if (has_not_been_filled[0][j])
        {
            fill_tool(rows-1, j, exterior_potential, exterior_fixed, has_not_been_filled);
        }   
    }    


}

File: src/BoundarySolver/BoundaryGenerationService.js




class BoundaryGenerationService {

    constructor(rows, cols){
  
        this.rows = rows;
        this.cols = cols;
  
        this.mesh = Array(rows).fill().map((() => Array(cols).fill(0)));
        this.fixedIndices = Array(rows).fill().map(() => (Array(cols).fill(false)));

        
    }

    placeRectangleBoundary(rectangle){

        let cornerX, cornerY, width, height, potential, fill;

        ({cornerX, cornerY, width, height, potential, fill} = {...rectangle} )

        if (fill){
            for(let i = cornerX; i <= cornerX+width; ++i){
                for(let j = cornerY; j <= cornerY+height; ++j){

                    if(this.coordInRange(i,j)){
                        this.mesh[i][j] = potential;
                        this.fixedIndices[i][j] = true;
                    }
                }
            }
        } else {

            for(let i = cornerX; i <= cornerX + width; ++i){

                if(this.coordInRange(i, cornerY)){
                    this.mesh[i][cornerY] = potential;
                    this.fixedIndices[i][cornerY] = true;
                }

                if(this.coordInRange(i, cornerY+height)){
                    this.mesh[i][cornerY+height] = potential;
                    this.fixedIndices[i][cornerY+height] = true;
                }
            }

            for(let j = cornerY; j <= cornerY + height; ++j){
                
                if(this.coordInRange(cornerX, j)){
                    this.mesh[cornerX][j] = potential;
                    this.fixedIndices[cornerX][j] = true;
                }

                if(this.coordInRange(cornerX+width, j)){
                    this.mesh[cornerX+width][j] = potential;
                    this.fixedIndices[cornerX+width][j] = true;
                }

            }

        }


    }




  
    placeCirclePotentialBoundary(circle){

        let centreX, centreY, radius, 
        setExternal, setInternal, 
        internalPotential, externalPotential, boundaryPotential;

        ({centreX, centreY, radius, 
            setExternal, setInternal, 
            internalPotential, externalPotential, boundaryPotential
        } = {...circle} )


        var hasNotBeenFilled = Array(this.rows).fill().map(() => Array(this.cols).fill(true));

        var allPassed = true;


        var xPos = 0;
        var yPos = radius;
        var decisionParam = 3 - (2 * radius);

        allPassed &= this.completeOctants(centreX, centreY, xPos, yPos, boundaryPotential, hasNotBeenFilled);

        while( yPos >= xPos ) {

            ++xPos;

            if (decisionParam > 0){

                --yPos;
                decisionParam = decisionParam + 4 * (xPos - yPos) + 10;
    
            } else {
    
                decisionParam = decisionParam + 4 * xPos + 6;
    
            }

            allPassed &= this.completeOctants(centreX, centreY, xPos, yPos, boundaryPotential, hasNotBeenFilled);

        }


        
        if(setInternal)
            this.fillTool(centreX, centreY, internalPotential, hasNotBeenFilled);

        if(setExternal)
            this.fillExterior(externalPotential, hasNotBeenFilled);
    
        return allPassed;
    };
  


    completeOctants(centreX, centreY, x, y, potential, hasNotBeenFilled){
        
        var allPassed = true

        allPassed &= this.fillInMeshForCircle(centreX + x, centreY + y, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX - x, centreY + y, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX + x, centreY - y, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX - x, centreY - y, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX + y, centreY + x, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX - y, centreY + x, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX + y, centreY - x, potential, hasNotBeenFilled);
        allPassed &= this.fillInMeshForCircle(centreX - y, centreY - x, potential, hasNotBeenFilled);

        return allPassed;

    }


    fillInMeshForCircle(x, y, potential, hasNotBeenFilled){
        
        if(!this.coordInRange(x,y))
            return false;
        
        this.mesh[x][y] = potential;
        this.fixedIndices[x][y] = true;
        hasNotBeenFilled[x][y] = false;

        return true;

    }



    fillTool(x, y, potential, hasNotBeenFilled){
        
        if(this.coordInRange(x,y) && hasNotBeenFilled[x][y]){

            this.fillInMeshForCircle(x, y, potential, hasNotBeenFilled);

            if(this.coordInRange(x+1, y))
            this.fillTool(x+1, y, potential, hasNotBeenFilled);
            
            if(this.coordInRange(x-1, y))
            this.fillTool(x-1, y, potential, hasNotBeenFilled);
            
            if(this.coordInRange(x, y+1))
            this.fillTool(x, y+1, potential, hasNotBeenFilled);
            
            if(this.coordInRange(x, y-1))
            this.fillTool(x, y-1, potential, hasNotBeenFilled);


        }

    }



    fillExterior(exteriorPotential, hasNotBeenFilled){

        for(let i = 0; i !== this.rows; ++i){
            if (hasNotBeenFilled[i][0])
            {
                this.fillTool(i, 0, exteriorPotential, hasNotBeenFilled);
            }   
        }
    
        //check right sie
        for(let j = 1; j !== this.cols; ++j){
                    if (hasNotBeenFilled[this.rows-1][j])
            {
                this.fillTool(this.rows-1, j, exteriorPotential, hasNotBeenFilled);
            }   
        }
    
        //check bottom side
        for(let i = 0; i !== this.rows-1; ++i){
            if (hasNotBeenFilled[i][this.cols-1])
            {
                this.fillTool(i, this.cols-1, exteriorPotential, hasNotBeenFilled);
            }   
        }
    
    
        //Check the left side
        for(let j = 1; j !== this.cols-1; ++j){
                    if (hasNotBeenFilled[0][j])
            {
                this.fillTool(this.rows-1, j, exteriorPotential, hasNotBeenFilled);
            }   
        }    
    }


    coordInRange(x, y){
        return x < this.rows && y < this.cols && x >= 0 && y >= 0;
    }




  }




export default BoundaryGenerationService;